src/bank.rs

```diff
@@ -501,22 +602,77 @@ impl Bank {
         }
     }
 
+    #[must_use]
+    pub fn process_and_record_transactions(
+        &self,
+        txs: &[Transaction],
+        poh: &PohRecorder,
+    ) -> Result<()> {
+        let now = Instant::now();
+        let locked_accounts = self.lock_accounts(txs);
+        let lock_time = now.elapsed();
+        let now = Instant::now();
+        let results = self.execute_and_commit_transactions(txs, locked_accounts);
+        let process_time = now.elapsed();
+        let now = Instant::now();
+        self.record_transactions(txs, results, poh)?;
```

Comment:
Instead of the early return on error here, why not call `unlock_accounts(txs)` first?  If if that's not the appropriate thing to do, should you hoist that call to `lock_accounts(tx)` to the caller? That asymmetry still bugs me. The caller needs to somehow know that a side-effect of calling this function is that it locks those accounts and it is the caller's responsibility to unlock them.

Reply:
@garious because recording and unlocking needs to be atomic.  as soon as you unlock some other thread could insert a conflicting entry into the ledger.

Reply:
That symmetry sounds ideal. It'd be helpful to understand more why that naive solution may not work for us.

Reply:
@garious i like that approach, doesn't seem to have any perf impact.  checkout the last commit and test